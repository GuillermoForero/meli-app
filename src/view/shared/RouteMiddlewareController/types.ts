export interface RouteMiddlewareControllerProps {
  middlewares: React.ReactElement<MiddlewareProps>[];
  children: React.ReactElement;
}

export interface MiddlewareProps {
  next?: () => void;
}
