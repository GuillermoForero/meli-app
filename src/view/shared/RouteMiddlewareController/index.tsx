import React, { useState, useCallback } from 'react';

import { RouteMiddlewareControllerProps } from 'view/shared/RouteMiddlewareController/types';
//The implementation of a kind of middleware and middleware controller is in general
//to execute actions before the component or the view is rendered and thus when it is rendered it has what it needs loaded
const RouteMiddlewareController: React.FC<RouteMiddlewareControllerProps> = ({
  children,
  middlewares,
}: RouteMiddlewareControllerProps) => {
  const middlewaresLength = middlewares.length;

  if (!middlewaresLength) {
    throw new Error('The middlewares array prop should not be empty');
  }
  const [currentMiddleware, setCurrentMiddleware] = useState(1);

  const next = useCallback(() => {
    setCurrentMiddleware(m => m + 1);
  }, []);

  const componentToRender =
    currentMiddleware <= middlewaresLength
      ? middlewares[currentMiddleware - 1]
      : children;

  return (
    <>
      {React.cloneElement(componentToRender, {
        next: currentMiddleware <= middlewaresLength ? next : undefined,
      })}
    </>
  );
};

export default RouteMiddlewareController;
