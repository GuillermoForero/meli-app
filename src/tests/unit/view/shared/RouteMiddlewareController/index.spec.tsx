import React from 'react';
import { render, fireEvent } from 'tests/support/customRender';
import RouteMiddlewareController from 'view/shared/RouteMiddlewareController';

const Middleware = ({ next }: { next?: () => void }) => {
  return <p onClick={next}>Middleware</p>;
};

const Component = () => {
  return <p>Component</p>;
};

class ErrorBoundary extends React.Component {
  state: any;
  constructor(props: any) {
    super(props);
    this.state = { hasError: false };
  }

  static getDerivedStateFromError() {
    return { hasError: true };
  }

  componentDidCatch() {}

  render() {
    if (this.state.hasError) {
      return <h1>Something went wrong.</h1>;
    }

    return this.props.children;
  }
}

describe('[RouteMiddlewareController]', () => {
  it('Renders a RouteMiddlewareController with a Middleware', () => {
    const { container, getByText, queryByText } = render(
      <RouteMiddlewareController middlewares={[<Middleware />]}>
        <Component />
      </RouteMiddlewareController>,
    );

    expect(getByText('Middleware')).toBeDefined();
    fireEvent.click(getByText('Middleware'));
    expect(queryByText('Middleware')).toBeNull();
    expect(getByText('Component')).toBeDefined();
    expect(container).toMatchSnapshot();
  });

  it('Renders a RouteMiddlewareController with multiple middlewares', () => {
    const { container, getByText, queryByText } = render(
      <RouteMiddlewareController
        middlewares={[<Middleware />, <Middleware />, <Middleware />]}
      >
        <Component />
      </RouteMiddlewareController>,
    );

    expect(getByText('Middleware')).toBeDefined();
    fireEvent.click(getByText('Middleware'));
    fireEvent.click(getByText('Middleware'));
    fireEvent.click(getByText('Middleware'));
    expect(queryByText('Middleware')).toBeNull();

    expect(getByText('Component')).toBeDefined();
    expect(container).toMatchSnapshot();
  });

  it('Renders a RouteMiddlewareController and throw an error when middlewares array is empty', () => {
    const { container, getByText } = render(
      <ErrorBoundary>
        <RouteMiddlewareController middlewares={[]}>
          <Component />
        </RouteMiddlewareController>
      </ErrorBoundary>,
    );

    expect(getByText('Something went wrong.')).toBeDefined();
    expect(container).toMatchSnapshot();
  });
});
