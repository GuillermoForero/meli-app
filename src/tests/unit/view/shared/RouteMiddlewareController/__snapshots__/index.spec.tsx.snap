// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[RouteMiddlewareController] Renders a RouteMiddlewareController and throw an error when middlewares array is empty 1`] = `
<div>
  <h1>
    Something went wrong.
  </h1>
</div>
`;

exports[`[RouteMiddlewareController] Renders a RouteMiddlewareController with a Middleware 1`] = `
<div>
  <p>
    Component
  </p>
</div>
`;

exports[`[RouteMiddlewareController] Renders a RouteMiddlewareController with multiple middlewares 1`] = `
<div>
  <p>
    Component
  </p>
</div>
`;
